### DISCUSSION TITLE
Cypress not interacting properly with DOM elements, even when the element is visible.

### DISCUSSION THREAD

**Discussion by yopdp** [cypress-io/cypress#27404] [Questions and Help]
**Alert checking function:**  

checkPopUpButton(alertType) {
    gen_info_elements
      .alertPopupBottom()
      .scrollIntoView()
      .then(($alert) => {
        if (alertType == true) {
          expect($alert).to.include.text(genInfoData.alertText.success);
        } else if (alertType == false) {
          expect($alert).to.include.text(genInfoData.alertText.failure);
        }
      });
  }"

**Alert is clearly visible and contains success/error alert text. But it is giving this kind of error message:**

expected <ion-toast#ion-overlay-3.ion-color.ion-color-danger.ios.toast-msg.hydrated> to contain text Try filling the mandatory fields., but the text was ''

**Comment by portyu9** [comment]
Hi @yopdp 

The important part of the error is that the element is an `<ion-toast>`.

Ionic's modern components use **Shadow DOM**, so the text you see on screen is not necessarily part of the host element's normal `textContent`. That explains why Cypress reports:

```text
the text was ''
```

even though the toast is visibly displaying the message.

Traverse into the toast's shadow root before asserting the text:

```js
const expectedText = alertType
  ? genInfoData.alertText.success
  : genInfoData.alertText.failure

gen_info_elements
  .alertPopupBottom()
  .should('be.visible')
  .shadow()
  .contains(expectedText)
  .should('be.visible')
```

You can also use Cypress's `includeShadowDom` support when querying elements inside Shadow DOM.

`scrollIntoView()` will not fix this particular problem. The host `<ion-toast>` can be completely visible while its displayed message still lives inside its shadow root.

I would also prefer a retryable Cypress assertion such as `.contains()`/`.should()` here rather than putting a synchronous `expect()` inside `.then()`, because a toast may still be rendering when the command first reaches it.

Please mark this answer as accepted if it helped, thank you.
